### PR TITLE
Fix slow test_moshika_greedy_unconditional_fp16

### LINKED ISSUE #38725: `MoshiIntegrationTests` started to fail after #34464
`MoshiIntegrationTests` started to fail after #34464.

### Reproduction

For example, 

> RUN_SLOW=1 python3 -m pytest -v tests/models/moshi/test_modeling_moshi.py::MoshiIntegrationTests::test_moshika_greedy_unconditional_fp16

Since then, during 8 months, there are several periods where those tests are failing with different errors (not even to run the forward/generate). But when they could run, the output values are consistent with the output values given by #34464. Only before/after #34464, the output values are different.

 @gante Could you check if the changed outputs are expected?

Your commit : 8a734ea2

Parent commit: 913330ca

(You will need to run `pip install -e .` when checking out to old commits like these)

### PR DESCRIPTION
**Pull request by manueldeprada**
Fix #38725

Coming from #38725, previously, e18f233f6c8cba029324e2868fb68abdaf6badf3 attempted to fix the default attention mask issue that appeated with #34464, but it was still failing the slow test `tests/models/moshi/test_modeling_moshi.py::MoshiIntegrationTests::test_moshika_greedy_unconditional_fp16`

History from git bisect:

- https://github.com/huggingface/transformers/blob/84a6789145c3d728f2e405d31e9a35df5d74f05c/src/transformers/modeling_utils.py#L1413 from #34873 broke the types for some commits, then it was fixed in a later commit.
- auto-compilation on generate also made it fail for some commits, it was also corrected after some commits.
- Then #35993 broke here the test (didnt have the -1):
https://github.com/huggingface/transformers/blob/36bf1d2324ee8a971763eabc2950a044d5eff3fa/src/transformers/generation/utils.py#L2090
- And #37873 modified sliding window behaviour breaking it as well (>= to >)
https://github.com/huggingface/transformers/blob/1b222903c3e1cfd9492d75e4b2548aa8bd458674/src/transformers/cache_utils.py#L1740

Setting cache_implementation="dynamic" makes the test pass, but the sliding window cache should not behave different. I believe this is due to the depth decoder being window 8 by default, but audio is confusing to me. 

This PR is not a fix: the modeling code should be changed to accomodate what I highlight in the diff

cc @eustlb @ydshieh

### CODE REVIEW

**Comment by manueldeprada** [comment]
run-slow: moshi

File: src/transformers/models/moshi/modeling_moshi.py [hunk 1]
```diff
@@ -1862,28 +1862,15 @@ def forward(
...
-        attention_mask_from_padding = input_ids.ne(pad_token_id).long()
-
-        attention_mask = (
-            attention_mask_from_padding * can_infer_attention_mask + default_attention_mask * ~can_infer_attention_mask
+        return super()._prepare_attention_mask_for_generation(
+            inputs_tensor=inputs_tensor,
+            generation_config=generation_config,
+            model_kwargs={},
         )
-        return attention_mask
```
Comment (eustlb) [inline src/transformers/models/moshi/modeling_moshi.py:1886]: can't we just remove _prepare_attention_mask_for_generation override here?

File: src/transformers/cache_utils.py [hunk 2]
```diff
@@ -95,7 +95,7 @@ def _sliding_cache_update(
     # Sliding window logic for generation phase or prefill < window
     slicing = torch.arange(max_cache_len, device=value_states.device)
     current_seq_len = cache_position[-1] + 1  # Use last position to determine current length
-    to_shift = current_seq_len > max_cache_len
+    to_shift = current_seq_len >= max_cache_len
```
Comment (eustlb) [inline src/transformers/cache_utils.py:98]: this will break other models no?

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -2481,7 +2481,7 @@ def generate(
         # - `model_kwargs` may be updated in place with a cache as defined by the parameters in `generation_config`.
         # - different models have a different cache name expected by the model (default = "past_key_values")
         # - `max_length`, prepared above, is used to determine the maximum cache length
-        max_cache_length = generation_config.max_length - 1
+        max_cache_length = generation_config.max_length
```
Comment (eustlb) [inline src/transformers/generation/utils.py:2484]: this will break other models no?

**Review by eustlb — COMMENTED** [review]
Thanks for the update @manueldeprada, still some work to come to fix moshi...
Is this PR still relevant concerning changes to `cache_utils.py` and `generation/utils.py `? If not can you solve merging conflicts?
